### PR TITLE
Replace more Unsafe.As usage with pointers

### PR DESCRIPTION
**Pull request by stephentoub**

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs [hunk 1]
```diff
@@ -652,7 +652,7 @@ public static void Write<T>(void* destination, T value)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static ref T AsRef<T>(void* source)
         {
-            return ref Unsafe.As<byte, T>(ref *(byte*)source);
+            return ref *(T*)source;
 
             // ldarg .0
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs:657]: Comments showing il disassembly and how methods are referenced in different runtimes can be deleted on the methods that you are touching.
```
        // CoreCLR:METHOD__UNSAFE__AS_REF_POINTER
        // AOT:AsRef
        // Mono:AsRef
```

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs [hunk 2]
```diff
@@ -652,7 +652,7 @@ public static void Write<T>(void* destination, T value)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static ref T AsRef<T>(void* source)
         {
-            return ref Unsafe.As<byte, T>(ref *(byte*)source);
+            return ref *(T*)source;
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs:655]: And also remove `[Intrinsic]`?

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/SafeBuffer.cs [hunk 3]
```diff
@@ -369,10 +369,11 @@ public ulong ByteLength
 
         private void SpaceCheck(byte* ptr, nuint sizeInBytes)
         {
-            if (_numBytes < sizeInBytes)
-                NotEnoughRoom();
-            if ((ulong)(ptr - (byte*)handle) > (_numBytes - sizeInBytes))
+            if ((_numBytes < sizeInBytes) ||
+                ((ulong)(ptr - (byte*)handle) > (_numBytes - sizeInBytes)))
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/SafeBuffer.cs:373]:
```suggestion
                ((nuint)(ptr - (byte*)handle) > (_numBytes - sizeInBytes)))
```

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs:655]: `[Intrinsic]` should be kept. It allows JIT to expand it as intrinsic that is more efficient than inlining the IL. Inlining the IL takes more time and creates extra IR that the optimizer needs to get rid of.

**Review by jkotas — APPROVED** [review]

**Comment by jkotas** [comment]
Some of the test failures are real problems - System.DataMisalignedException : A datatype misalignment was detected in a load or store instruction.

**Comment by stephentoub** [comment]
I reverted the changes to SafeBuffer.

**Pull request merged**